### PR TITLE
BUG: Incorrect index shape when using a user-defined function for aggregating a grouped series with object-typed index.

### LINKED ISSUE #40014: BUG: "groupby" failed for dataframe with object-type index
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

I have a dataframe where the index is object type, and I found the error by accident. The following code is a simplified reproduction of the error.

```python
df = pd.DataFrame(
    { "c0": ["x", "x", "x"],
      "c1": ["x", "x", "y"],
      "p": [0,1,2] } )
​
df.index = df.index.astype("O")
df.groupby(['c0','c1']).p.agg(lambda x:all(x>0))
```
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-227-54707938a409> in <module>
      5 
      6 df.index = df.index.astype("O")
----> 7 df.groupby(['c0','c1']).p.agg(lambda x:all(x>0))

/opt/conda/lib/python3.8/site-packages/pandas/core/groupby/generic.py in aggregate(self, func, engine, engine_kwargs, *args, **kwargs)
    254 
    255             if self.grouper.nkeys > 1:
--> 256                 return self._python_agg_general(func, *args, **kwargs)
    257 
    258             try:

/opt/conda/lib/python3.8/site-packages/pandas/core/groupby/groupby.py in _python_agg_general(self, func, *args, **kwargs)
   1171             try:
   1172                 # if this function is invalid for this dtype, we will ignore it.
-> 1173                 result, counts = self.grouper.agg_series(obj, f)
   1174             except TypeError:
   1175                 continue

/opt/conda/lib/python3.8/site-packages/pandas/core/groupby/ops.py in agg_series(self, obj, func)
    704 
    705         try:
--> 706             return self._aggregate_series_fast(obj, func)
    707         except ValueError as err:
    708             if "Must produce aggregated value" in str(err):

/opt/conda/lib/python3.8/site-packages/pandas/core/groupby/ops.py in _aggregate_series_fast(self, obj, func)
    730         group_index = algorithms.take_nd(group_index, indexer, allow_fill=False)
    731         grouper = libreduction.SeriesGrouper(obj, func, group_index, ngroups, dummy)
--> 732         result, counts = grouper.get_result()
    733         return result, counts
    734 

pandas/_libs/reduction.pyx in pandas._libs.reduction.SeriesGrouper.get_result()

pandas/_libs/reduction.pyx in pandas._libs.reduction._BaseGrouper._apply_to_group()

/opt/conda/lib/python3.8/site-packages/pandas/core/groupby/groupby.py in <lambda>(x)
   1158     def _python_agg_general(self, func, *args, **kwargs):
   1159         func = self._is_builtin_func(func)
-> 1160         f = lambda x: func(x, *args, **kwargs)
   1161 
   1162         # iterate through "columns" ex exclusions to populate output dict

<ipython-input-227-54707938a409> in <lambda>(x)
      5 
      6 df.index = df.index.astype("O")
----> 7 df.groupby(['c0','c1']).p.agg(lambda x:all(x>0))

/opt/conda/lib/python3.8/site-packages/pandas/core/ops/common.py in new_method(self, other)
     63         other = item_from_zerodim(other)
     64 
---> 65         return method(self, other)
     66 
     67     return new_method

/opt/conda/lib/python3.8/site-packages/pandas/core/arraylike.py in __gt__(self, other)
     43     @unpack_zerodim_and_defer("__gt__")
     44     def __gt__(self, other):
---> 45         return self._cmp_method(other, operator.gt)
     46 
     47     @unpack_zerodim_and_defer("__ge__")

/opt/conda/lib/python3.8/site-packages/pandas/core/series.py in _cmp_method(self, other, op)
   4975         res_values = ops.comparison_op(lvalues, rvalues, op)
   4976 
-> 4977         return self._construct_result(res_values, name=res_name)
   4978 
   4979     def _logical_method(self, other, op):

/opt/conda/lib/python3.8/site-packages/pandas/core/series.py in _construct_result(self, result, name)
   2759         # We do not pass dtype to ensure that the Series constructor
   2760         #  does inference in the case where `result` has object-dtype.
-> 2761         out = self._constructor(result, index=self.index)
   2762         out = out.__finalize__(self)
   2763 

/opt/conda/lib/python3.8/site-packages/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    346                 try:
    347                     if len(index) != len(data):
--> 348                         raise ValueError(
    349                             f"Length of passed values is {len(data)}, "
    350                             f"index implies {len(index)}."

ValueError: Length of passed values is 1, index implies 2.
```

#### Problem description

If the index is not an object type, you will not get this error, so I guess there is something wrong with object type index.

#### Expected Output

```python
c0  c1
x   x     False
    y      True
Name: p, dtype: bool
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.121-linuxkit
Version          : #1 SMP Tue Dec 1 17:50:32 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.2
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.6.0.post20200917
Cython           : 0.29.21
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.2
bottleneck       : 1.3.2
fsspec           : 0.8.3
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : 1.0.1
pyxlsb           : None
s3fs             : 0.2.2
scipy            : 1.5.2
sqlalchemy       : 1.3.19
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : None
numba            : 0.51.2
</details>

### PR DESCRIPTION
**Pull request by DriesSchaumont**
- [x] closes #40014
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_function.py [hunk 1]
```diff
@@ -1087,3 +1087,11 @@ def test_groupby_sum_below_mincount_nullable_integer():
     result = grouped.sum(min_count=2)
     expected = DataFrame({"b": [pd.NA] * 3, "c": [pd.NA] * 3}, dtype="Int64", index=idx)
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_index_object_dtype():
```
Comment (mzeitlin11) [inline pandas/tests/groupby/test_function.py:1092]: I think `pandas/tests/groupby/aggregate/test_aggregate.py` might be a better place for this test - looks like there are some pretty similar ones

File: pandas/tests/groupby/test_function.py [hunk 2]
```diff
@@ -1087,3 +1087,11 @@ def test_groupby_sum_below_mincount_nullable_integer():
...
+    # GH 40014
+    df = DataFrame({"c0": ["x", "x", "x"], "c1": ["x", "x", "y"], "p": [0, 1, 2]})
+    df.index = df.index.astype("O")
+    grouped = df.groupby(["c0", "c1"])
+    grouped.p.agg(lambda x: all(x > 0))
```
Comment (mzeitlin11) [inline pandas/tests/groupby/test_function.py:1097]: Even though this test does verify no error is raised, may as well also follow result/expected pattern and verify correct output

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -738,7 +738,7 @@ def agg_series(self, obj: Series, func: F):
             # TODO: can we get a performant workaround for EAs backed by ndarray?
             return self._aggregate_series_pure_python(obj, func)
 
-        elif obj.index._has_complex_internals:
+        elif obj.index._has_complex_internals or obj.index.dtype == "object":
```
Comment (mzeitlin11) [inline pandas/core/groupby/ops.py:741]: A comment here might be helpful explaining why we should expect an object type index to raise in the fastpath

**Review by mzeitlin11 — COMMENTED** [review]
Thanks for the pr! Some quick comments, plus can you please add a whatsnew?

File: pandas/core/groupby/ops.py (hunk 3 above)
Comment (DriesSchaumont) [reply pandas/core/groupby/ops.py:741]: Yes, definitely! Done

File: pandas/tests/groupby/test_function.py (hunk 1 above)
Comment (DriesSchaumont) [reply pandas/tests/groupby/test_function.py]: Done!

File: pandas/tests/groupby/test_function.py (hunk 2 above)
Comment (DriesSchaumont) [reply pandas/tests/groupby/test_function.py]: Done!

**Comment by DriesSchaumont** [comment]
> Thanks for the pr! Some quick comments, plus can you please add a whatsnew?

Done. Added whatsnew  targetting 1.2.4, but let me know if this is not the case.

File: doc/source/whatsnew/v1.2.4.rst [hunk 4]
```diff
@@ -29,7 +29,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 
--
+- Bug in groupby aggregation when using an object-typed index (issue:`40014`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.4.rst:32]: we are rarely going to target anything but master 
move to 1.3

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.2.4.rst (hunk 4 above)
Comment (DriesSchaumont) [reply doc/source/whatsnew/v1.2.4.rst]: Good to know. Updated

File: doc/source/whatsnew/v1.3.0.rst [hunk 5]
```diff
@@ -780,6 +780,7 @@ Groupby/resample/rolling
 - :class:`core.window.ewm.ExponentialMovingWindow` now raises a ``NotImplementedError`` when specifying ``times`` with ``adjust=False`` due to an incorrect calculation (:issue:`40098`)
 - Bug in :meth:`Series.asfreq` and :meth:`DataFrame.asfreq` dropping rows when the index is not sorted (:issue:`39805`)
 - Bug in aggregation functions for :class:`DataFrame` not respecting ``numeric_only`` argument when ``level`` keyword was given (:issue:`40660`)
+- Bug in groupby aggregation when using an object-typed index (issue:`40014`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:783]: either use ```:class:`Index` ``` or Index

**Review by jreback — CHANGES_REQUESTED** [review]
small comment

File: doc/source/whatsnew/v1.3.0.rst (hunk 5 above)
Comment (DriesSchaumont) [reply doc/source/whatsnew/v1.3.0.rst]: Adjusted!

File: pandas/core/groupby/ops.py [hunk 6]
```diff
@@ -738,8 +738,12 @@ def agg_series(self, obj: Series, func: F):
...
             return self._aggregate_series_pure_python(obj, func)
 
-        elif obj.index._has_complex_internals:
-            # Preempt TypeError in _aggregate_series_fast
+        elif obj.index._has_complex_internals or obj.index.dtype == "object":
+            # (complex internals): Preempt TypeError in _aggregate_series_fast
+            # (object index dtype): _aggregate_series_fast raises TypeError
+            # when applying func because the group indiced become malformatted:
+            # correct indices are in group.index._index_data, but not in
+            # group.index._data.
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:746]: On master, replacing the lamba in the linked issue with `lambda x: 1` does not raise and gives the correct answer. It seems to me that just having an object dtype is not root cause here, and we will be hurting performance for this special case.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (rhshadrach) [reply pandas/core/groupby/ops.py:746]: https://github.com/pandas-dev/pandas/issues/40014#issuecomment-817196445

Comment (DriesSchaumont) [reply pandas/core/groupby/ops.py:746]: I see. I will keep looking for the root cause and try to provide a fix. I will mark this PR as a draft in the meantime.

**Comment by DriesSchaumont** [comment]
I am opening this PR for review again. I have added another test, not sure on which one to keep. Open to any suggestions.

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (DriesSchaumont) [reply pandas/core/groupby/ops.py]: @rhshadrach I have updated this PR.

File: doc/source/whatsnew/v1.3.0.rst [hunk 7]
```diff
@@ -781,6 +781,7 @@ Groupby/resample/rolling
 - :class:`core.window.ewm.ExponentialMovingWindow` now raises a ``NotImplementedError`` when specifying ``times`` with ``adjust=False`` due to an incorrect calculation (:issue:`40098`)
 - Bug in :meth:`Series.asfreq` and :meth:`DataFrame.asfreq` dropping rows when the index is not sorted (:issue:`39805`)
 - Bug in aggregation functions for :class:`DataFrame` not respecting ``numeric_only`` argument when ``level`` keyword was given (:issue:`40660`)
+- Bug in groupby aggregation when using an object-typed :class:`Index` (issue:`40014`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:784]: can you be more specific here? as a user i am not sure what i am to do with this statement

Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:784]: add a 'for example df.groupby(....)......` or similar, in particular this is only for a user supplied function (e.g. via .apply right)?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 8]
```diff
@@ -1221,3 +1221,16 @@ def test_aggregate_numeric_object_dtype():
         {"key": ["A", "B"], "col1": ["a", "c"], "col2": [0, 2]}
     ).set_index("key")
     tm.assert_frame_equal(result, expected)
+
+
+def test_groupby_index_object_dtype():
+    # GH 40014
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1227]: can you add a 1-liner comment about the expected index here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 7 above)
Comment (DriesSchaumont) [reply doc/source/whatsnew/v1.3.0.rst]: Adjusted, hope this is more clear?

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 8 above)
Comment (DriesSchaumont) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1227]: I wasnt quite not sure what you meant/expected, but I added a comment. Let me know if I misinterpreted something.

File: pandas/tests/groupby/test_bin_groupby.py [hunk 9]
```diff
@@ -27,6 +27,21 @@ def test_series_grouper():
...
+def test_series_grouper_result_length_difference():
+    obj = Series(np.random.randn(10), dtype="float64")
+    obj.index = obj.index.astype("O")
+    labels = np.array([-1, -1, -1, 0, 0, 0, 1, 1, 1, 1], dtype=np.intp)
+
+    grouper = libreduction.SeriesGrouper(obj, lambda x: all(x > 0), labels, 2)
+    result, counts = grouper.get_result()
+
+    expected = np.array([all(obj[3:6] > 0), all(obj[6:] > 0)])
+    tm.assert_almost_equal(result, expected)
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_bin_groupby.py:39]: Why use almost_equal here?

File: pandas/tests/groupby/test_bin_groupby.py [hunk 10]
```diff
@@ -27,6 +27,21 @@ def test_series_grouper():
     tm.assert_almost_equal(counts, exp_counts)
 
 
+def test_series_grouper_result_length_difference():
+    obj = Series(np.random.randn(10), dtype="float64")
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_bin_groupby.py:31]: Issue number

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This looks great! Can you just check on my example in the issue and report back? Do you understand why the "> 0" was causing it to fail? It's still not clear to me from the diff in this PR.

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (DriesSchaumont) [reply pandas/core/groupby/ops.py]: > 
> 
> This looks great! Can you just check on my example in the issue and report back? Do you understand why the "> 0" was causing it to fail? It's still not clear to me from the diff in this PR.

I think what happens here is that `x.gt(0)` calls `_cmp_method` and constructs a new `Series` through `_construct_result`. Note that the index length check happens when a new `Series` is constructed. `x.sum()` on the other hand does not construct a new series. The return type is `numpy.int64`. So it seems that using any function that constructs a new `Series` will fail, while the other will not. But the index state in the user-defined function is always corrupt.
A good test would be to use 

```python3
import pandas.core.common as com

def foo(x):
    com.require_length_match(x.values, x.index)
    return x
```

File: pandas/tests/groupby/test_bin_groupby.py (hunk 9 above)
Comment (DriesSchaumont) [reply pandas/tests/groupby/test_bin_groupby.py]: It is not necessary. I copied it from the test above. Adjusted.

File: pandas/tests/groupby/test_bin_groupby.py (hunk 10 above)
Comment (DriesSchaumont) [reply pandas/tests/groupby/test_bin_groupby.py:32]: Added.

**Comment by DriesSchaumont** [comment]
> 
> 
> This looks great! Can you just check on my example in the issue and report back? Do you understand why the "> 0" was causing it to fail? It's still not clear to me from the diff in this PR.

I addressed your question here: https://github.com/pandas-dev/pandas/issues/40014#issuecomment-819489220

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. cc @jbrockmendel if any comments

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
@rhshadrach over to you

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @DriesSchaumont, very nice!

**Pull request merged**